File: clients/ts/src/evm/finalize.ts
import { decodeEventLog } from "viem";
import {
  publicClient,
  walletClient,
  CAMPAIGN_ADDRESS,
  CAMPAIGN_ABI,
} from "./config.js";
import { printResult, printError } from "../shared/output.js";

async function main() {
  // Read campaign deadline and advance time past it (Hardhat localnet only)
  const deadline = (await publicClient.readContract({
    address: CAMPAIGN_ADDRESS,
    abi: CAMPAIGN_ABI,
    functionName: "deadline",
  })) as bigint;

  const block = await publicClient.getBlock();
  if (block.timestamp <= deadline) {
    const secondsToAdvance = Number(deadline - block.timestamp) + 1;
    await publicClient.request({
      method: "evm_increaseTime" as any,
      params: [secondsToAdvance] as any,
    });
    await publicClient.request({
      method: "evm_mine" as any,
      params: [] as any,
    });
  }

  const start = performance.now();

  const hash = await walletClient.writeContract({
    address: CAMPAIGN_ADDRESS,
    abi: CAMPAIGN_ABI,
    functionName: "finalize",
  });

  const receipt = await publicClient.waitForTransactionReceipt({ hash });
  const elapsed = performance.now() - start;

  // Parse Finalized event
  let successful: boolean | null = null;
  let totalRaised = "";
  for (const log of receipt.logs) {
    if (log.address.toLowerCase() === CAMPAIGN_ADDRESS.toLowerCase()) {
      try {
        const event = decodeEventLog({
          abi: CAMPAIGN_ABI,
          data: log.data,
          topics: log.topics,
        });
        if (event.eventName === "Finalized") {
          const args = event.args as unknown as { successful: boolean; totalRaised: bigint };
          successful = args.successful;
          totalRaised = args.totalRaised.toString();
          break;
        }
      } catch {}
    }
  }

  printResult({
    chain: "evm",
    operation: "finalize",
    txHash: hash,
    blockNumber: Number(receipt.blockNumber),
    gasUsed: Number(receipt.gasUsed),
    status: receipt.status === "success" ? "success" : "reverted",
    timestamp: new Date().toISOString(),
    elapsedMs: Math.round(elapsed),
    data: { successful, totalRaised },
  });
}

main().catch((err) => printError("finalize", err));

File: clients/python/evm/finalize.py
"""
evm/finalize.py -- Finalize a crowdfunding campaign.

Usage:
    python -m clients.python evm:finalize
"""

from __future__ import annotations

import argparse
import sys
import time

from clients.python.evm import config as evm_config
from clients.python.evm.client import get_web3, load_abi, get_account, build_and_send
from clients.python.shared.output import TxOutput, print_result, print_error, _now_iso


def finalize(variant: str) -> TxOutput:
    """Call finalize() on the campaign contract."""
    w3 = get_web3()
    if not w3.is_connected():
        raise ConnectionError(f"Cannot connect to {evm_config.RPC_URL}")

    account = get_account()

    campaign_addr = evm_config.CAMPAIGN_ADDRESS
    if not campaign_addr:
        raise EnvironmentError("CAMPAIGN_ADDRESS env var is not set.")

    _, campaign_artifact_path, _ = evm_config.EVM_VARIANT_ARTIFACTS[variant]
    campaign_abi = load_abi(campaign_artifact_path)
    campaign = w3.eth.contract(address=campaign_addr, abi=campaign_abi)

    receipt, latency = build_and_send(w3, campaign.functions.finalize(), account, gas=200_000)

    return TxOutput(
        chain="evm",
        operation="finalize",
        tx_hash=receipt["transactionHash"].hex(),
        block_number=receipt["blockNumber"],
        gas_used=receipt["gasUsed"],
        status="success" if receipt["status"] == 1 else "reverted",
        timestamp=_now_iso(),
        elapsed_ms=latency,
        data={"variant": variant},
    )


def main(args=None) -> None:
    parser = argparse.ArgumentParser(description="Finalize a campaign.")
    parser.add_argument("--variant", default=evm_config.VARIANT)
    parsed = parser.parse_args(args)

    try:
        output = finalize(parsed.variant)
        print_result(output)
    except Exception as exc:
        print_error("finalize", exc, chain="evm")


if __name__ == "__main__":
    main()
